File: go-protoactor/internal/models/params.go
package models

import (
	"benchmarking/chatapp/actor/internal/protos/messages"
	"fmt"
	"os"
	"time"

	"gopkg.in/yaml.v3"
)

type TestType string

const (
	Throughput    TestType = "throughput"
	Reconnection  TestType = "reconnection_time"
	DetectionTime TestType = "detection_time"
)

// String returns the string representation of TestType
func (t TestType) String() string {
	return string(t)
}

// UnmarshalYAML ensures only valid enum values are accepted for TestType
func (t *TestType) UnmarshalYAML(value *yaml.Node) error {
	var s string
	if err := value.Decode(&s); err != nil {
		return err
	}

	switch s {
	case string(Throughput), string(Reconnection), string(DetectionTime):
		*t = TestType(s)
		return nil
	default:
		return fmt.Errorf("invalid test_type: %s", s)
	}
}

func (t TestType) ToProto() messages.TestType {
	switch t {
	case Throughput:
		return messages.TestType_THROUGHPUT
	case Reconnection:
		return messages.TestType_RECONNECTION
	case DetectionTime:
		return messages.TestType_DETECTION
	default:
		panic("error on TestType")
	}
}

type Params struct {
	TestType         TestType `yaml:"test_type"`
	TestDurationSecs int      `yaml:"test_duration_seconds"`
	NumSupervisor    int      `yaml:"num_supervisor"`
	ChatsPerSup      int      `yaml:"chats_per_sup"`
	ClientsPerServer int      `yaml:"clients_per_server"`
	MsgType          string   `yaml:"msg_type"`
	FaultPauseMs     int      `yaml:"fault_pause_ms"`
}

type Config struct {
	Params Params `yaml:"params"`
}

func (params *Params) Print(testId string) {
	fmt.Println("################## TEST [Go] ##################")
	fmt.Printf("Test id: %s\n", testId)
	fmt.Printf("Test type: %s\n", params.TestType)
	fmt.Printf("Test duration: %d seconds\n", params.TestDurationSecs)
	fmt.Printf("Number of supervisors: %d\n", params.NumSupervisor)
	fmt.Printf("Chats per supervisor: %d\n", params.ChatsPerSup)
	fmt.Printf("Clients per server: %d\n", params.ClientsPerServer)
	fmt.Printf("Message type: %s\n", params.MsgType)
	fmt.Printf("Fault pause (ms): %d\n", params.FaultPauseMs)
	fmt.Println("###################################################")
}

func LoadParams(filename string) (*Params, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	config := &Config{}
	if err := yaml.Unmarshal(data, config); err != nil {
		return nil, err
	}
	return &config.Params, nil
}

type TestInfo struct {
	InitTime time.Time
	EndTime  time.Time
}

File: elixir/lib/models/params.ex
defmodule Models.FaultMessageType do
  @type t :: :cpu_intensive | :ram_intensive | :error | :random | :none

  @spec from_string(String.t()) :: t()
  def from_string(value) do
    case value do
      "cpu_intensive" -> :cpu_intensive
      "ram_intensive" -> :ram_intensive
      "error" -> :error
      "random" -> :random
      "none" -> :none
      _ -> raise ArgumentError, "Invalid FaultMessageType: #{value}"
    end
  end

  @spec to_string(t()) :: String.t()
  def to_string(type) do
    case type do
      :cpu_intensive -> "cpu_intensive"
      :ram_intensive -> "ram_intensive"
      :error -> "error"
      :random -> "random"
      :none -> "none"
    end
  end
end

defmodule Models.TestType do
  @type t :: :throughput | :reconnection_time | :detection_time

  @spec from_string(String.t()) :: t()
  def from_string(value) do
    case value do
      "throughput" -> :throughput
      "reconnection_time" -> :reconnection_time
      "detection_time" -> :detection_time
      _ -> raise ArgumentError, "Invalid TestType: #{value}"
    end
  end

  @spec to_string(t()) :: String.t()
  def to_string(type) do
    case type do
      :throughput -> "throughput"
      :reconnection_time -> "reconnection_time"
      :detection_time -> "detection_time"
    end
  end
end

defmodule Models.Params do
  @enforce_keys [
    :test_type,
    :test_duration_seconds,
    :num_supervisor,
    :chats_per_sup,
    :clients_per_server,
    :msg_type,
    :fault_pause_ms,
    :client_base_rate,
    :client_ceil_rate
  ]

  defstruct @enforce_keys

  @type t :: %__MODULE__{
          test_type: Models.TestType.t(),
          test_duration_seconds: integer(),
          num_supervisor: integer(),
          chats_per_sup: integer(),
          clients_per_server: integer(),
          msg_type: Models.FaultMessageType.t(),
          fault_pause_ms: integer(),
          client_base_rate: integer(),
          client_ceil_rate: integer()
        }

  @spec new(
          test_type :: Models.TestType.t(),
          test_duration_seconds :: integer(),
          num_supervisor :: integer(),
          chats_per_sup :: integer(),
          clients_per_server :: integer(),
          msg_type :: Models.FaultMessageType.t() | String.t(),
          fault_pause_ms :: integer(),
          client_base_rate :: integer(),
          client_ceil_rate :: integer()
        ) :: t()
  def new(
        test_type,
        test_duration_seconds,
        num_supervisor,
        chats_per_sup,
        clients_per_server,
        msg_type,
        fault_pause_ms,
        client_base_rate,
        client_ceil_rate
      ) do
    msg_type =
      if is_binary(msg_type), do: Models.FaultMessageType.from_string(msg_type), else: msg_type

    test_type =
      if is_binary(test_type), do: Models.TestType.from_string(test_type), else: test_type

    %__MODULE__{
      test_type: test_type,
      test_duration_seconds: test_duration_seconds,
      num_supervisor: num_supervisor,
      chats_per_sup: chats_per_sup,
      clients_per_server: clients_per_server,
      msg_type: msg_type,
      fault_pause_ms: fault_pause_ms,
      client_base_rate: client_base_rate,
      client_ceil_rate: client_ceil_rate
    }
  end
end
